### ISSUE TITLE
`modernize-use-emplace` not reported on usage via accessor

### ISSUE THREAD

**Issue by firewave** [llvm/llvm-project#56721]
```cpp
#include <stack>
#include <string>
#include <vector>

class VariableMap {
private:
    std::stack<std::vector<std::pair<std::string, int>>> mScopeInfo;
public:
    void addVariable(const std::string& varname);
};

void VariableMap::addVariable(const std::string& varname)
{
    mScopeInfo.top().push_back(std::pair<std::string, int>(varname, 0));
}
```
https://godbolt.org/z/KGdYxsoY4

If you remove the `std::stack` wrapper and access it directly it is being detected.

**Comment by firewave** [comment]
This appears to be related to it being a STL container. Using an accessor with the object directly available reports the issue:

```cpp
#include <stack>
#include <string>
#include <vector>

class VariableMap {
private:
    std::vector<std::pair<std::string, int>> mScopeInfo;

    std::vector<std::pair<std::string, int>>& scopeInfo()
    {
        return mScopeInfo;
    }
public:
    void addVariable(const std::string& varname);
};

void VariableMap::addVariable(const std::string& varname)
{
   scopeInfo().push_back(std::pair<std::string, int>(varname, 0));
}
```

```
<source>:19:16: warning: use emplace_back instead of push_back [modernize-use-emplace]
   scopeInfo().push_back(std::pair<std::string, int>(varname, 0));
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~          ~
               emplace_back(
```
https://godbolt.org/z/nhT8YGsqT

**Comment by firewave** [comment]
This also prevents the `unnecessary temporary object` version of this check being detected:

```cpp
#include <stack>
#include <string>
#include <vector>

class VariableMap {
private:
    std::stack<std::vector<std::pair<std::string, int>>> mScopeInfo;
    std::vector<std::pair<std::string, int>> mScopeInfo2;
public:
    void addVariable(const std::string& varname);
    void addVariable2(const std::string& varname);
};

void VariableMap::addVariable(const std::string& varname)
{
    mScopeInfo.top().emplace_back(std::pair<std::string, int>(varname, 0)); // no warning
}

void VariableMap::addVariable2(const std::string& varname)
{
    mScopeInfo2.emplace_back(std::pair<std::string, int>(varname, 0)); // warning
}
```
https://godbolt.org/z/ao3dxx5hn

**Comment by smcpeak** [comment]
The bug relates to how `reference` is declared in the container.  That in turn causes `clang-tidy` to not realize that the return value of the accessor is an instance of a relevant container.

The core of `UseEmplaceCheck.cpp` is:

```
  auto CallPushBack = cxxMemberCallExpr(
      hasDeclaration(functionDecl(hasName("push_back"))),
      on(hasType(cxxRecordDecl(hasAnyName(SmallVector<StringRef, 5>(
          ContainersWithPushBack.begin(), ContainersWithPushBack.end()))))));
```

Using the example from #57392, the following simplified matcher will report the non-nested case, as expected:

```
  cxxMemberCallExpr(
    hasDeclaration(functionDecl(hasName("push_back"))),
    on(hasType(cxxRecordDecl(hasName("std::vector"))))
  )
```

Reports:
```
  vector<pair<int, int>> a;
  a.push_back(make_pair(1, 2)); //caught
```

However, to get it to match the nested case (when compiling with GCC-9.3.0 headers), it must be changed to:

```
  cxxMemberCallExpr(
    hasDeclaration(functionDecl(hasName("push_back"))),
    on(hasType(namedDecl(hasName("::__gnu_cxx::__alloc_traits<std::allocator<std::vector<std::pair<int, int>>>, std::vector<std::pair<int, int>>>::value_type"))))
  )
```

Reports:
```
  vector<vector<pair<int, int>>> b(1);
  b[0].push_back(make_pair(1, 2)); //not caught
```

That is, the type is no longer a `cxxRecordDecl`, and its name is a long template-id.  That template-id is derived from the code used to declare `std::vector::reference` in the headers.  It should be semantically equivalent to `std::pair<int, int>`, but that equivalence is evidently obscured from `clang-tidy` here.

**Comment by firewave** [comment]
@BigPeet given your recent submissions - could you take a look at this as well?

**Comment by BigPeet** [comment]
Yes, I can take a look at this in the next couple of days.

**Comment by firewave** [comment]
Nice. This is very appreciated.

**Comment by njames93** [comment]
Take a look at desugared types, this should match what's needed

**Comment by firewave** [comment]
Turns out this is already being detected:
```
<source>:14:22: warning: use emplace_back instead of push_back [modernize-use-emplace]
    mScopeInfo.top().push_back(std::pair<std::string, int>(varname, 0));
                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~          ~
                     emplace_back(
```

**Comment by firewave** [comment]
This is fixed with clang-tidy 16.
